solutions/cf-888G.mdx

```diff
@@ -0,0 +1,160 @@
+
+Because $b$ is the first different bit and because the array is sorted, $L$ contains some prefix of the subarray and $R$ contains some suffix. That is to say, $L = A[l:m - 1]$ and $R = [m : r]$ for some $l < m \leq r$. 
+
+**Lemma:** In a minimum xor spanning tree, there is exactly 1 edge between an element in $L$ and an element in $R$.
+
+**Proof:** By definition, a spanning tree must connect all vertices, which implies the existence of at least 1 edge between $L$ and $R$. Assume there is two edges between $L$ and $R$, with endpoints $(l_1, r_1)$ and $(l_2, r_2)$ respectively, where $l_1, l_2 \in L$ and $r_1, r_2 \in R$. The tree is constructed in one of the following ways, where solid lines represent edges and dashed lines represent paths. 
+
+![](<cf-888G/double1.png>)
+
+![](<cf-888G/double2.png>)
+
+Without loss of generality, disconnect $(l_2, r_2)$. The spanning tree becomes disconnected, with either $l_2$ or $r_2$ in the component that _does not_ contain $l_1$ and $r_1$. It is always more optimal to connect this node to the corresponding edge node. That is, if $l_2$ is not in the same component as $l_1$, then an edge between $l_1$ and $l_2$ will always be better than one between $l_2$ and $r_2$. 
+
+![](<cf-888G/single.png>)
+
+By definition, $l_1$ and $l_2$ share a bit $b$, so the xor of their values does not contain bit $b$, whereas $l_2 \texttt{ xor } r_2$ always contains $b$. 
+Note that this implies that it is always better to connect $l_2$ to $l_1$; however, this configuration may not be optimal. The proof is analguous if instead $r_2$ is disconnected. $\blacksquare$
+
+Thus, the divide and conquer algorithm will proceed as follows: We will partition the subarray into $L$ and $R$. This step takes $\mathcal{O}(n)$ time. Note that if either $L$ or $R$ is empty (if all numbers in the range contain $b$ or if none of them contain $b$), we immeditately recurse with the same subarray and $b - 1$ as the bit. Next, we will choose an element from $L$ and an element $R$ with a minimum pairwise xor, and add this to the answer. Lastly, recursively run the algorithm on $L$ and $R$ with bit $b - 1$. 
+
+How do we find two values such that their xor is minimized? One way to do this efficiently is with a trie. In the trie, each number is represented as a string of its binary representation, starting with the most significant digit and ending with the least. For xor queries, we can do a greedy traversal from root to leaf. When a transition exists that corresponds to the current bit of the query number, we greedily take that transition; otherwise, we are forced to take the unoptimal transition, and add the weight of that bit to the answer. Because there are $n$ strings and each string is $\log n$ long, this step takes $\mathcal{O}(n \log n)$.
```

Comment:
> and each string is $\log n$ long,

This isn't true? It would be more accurate to use a different variable to denote the number of bits in each vertex

Reply:
don't rlly feel like reviewing the rest, it's probably fine

Reply:
> This isn't true? It would be more accurate to use a different variable to denote the number of bits in each vertex (ex. B in your solution)

Wouldn't this value just be $log a$ ?

Reply:
I might be misunderstanding what you mean